pkg/watch/watch.go

```diff
@@ -178,34 +187,43 @@ func (o *WatchClient) WatchAndPush(out io.Writer, parameters WatchParameters) er
 	// mutex as they are shared between goroutines to communicate
 	// sync state/events.
 	var (
-		changeLock   sync.Mutex
-		dirty        bool
-		lastChange   time.Time
-		watchError   error
-		deletedPaths []string
-		changedFiles []string
+		lastChange                     time.Time
+		watchError                     error
+		deletedPaths                   []string
+		changedFiles                   []string
+		hasFirstSuccessfulPushOccurred bool
+		wg                             sync.WaitGroup
+		showWaitingMessage             bool
+		delay                          time.Duration
 	)
 
+	delay = 1 * time.Second
+
 	watcher, err := fsnotify.NewWatcher()
 	if err != nil {
 		return fmt.Errorf("error setting up filesystem watcher: %v", err)
 	}
 	defer watcher.Close()
 	defer close(parameters.ExtChan)
 
+	// adding watch on the root folder and the sub folders recursively
+	// so directory and the path in addRecursiveWatch() are the same
+	err = addRecursiveWatch(watcher, parameters.Path, parameters.FileIgnores)
+	if err != nil {
+		return fmt.Errorf("error watching source path %s: %v", parameters.Path, err)
+	}
+
+	log.Finfof(out, "\nWatching for changes in the current directory %s\n\nPress Ctrl+c to exit.", parameters.Path)
 	// This goroutine listens for either file change events from fsnotify, fs errors, or a terminate signal
 	// The results are stored in the variables defined in the var( ... ) block above
-	go func() {
-		for {
-			select {
-			case extMsg := <-parameters.ExtChan:
-				if extMsg {
-					changeLock.Lock()
-					watchError = ErrUserRequestedWatchExit
-					changeLock.Unlock()
-				}
-			case event := <-watcher.Events:
-				changeLock.Lock()
+	wg.Add(1)
+	for {
+		select {
+		case event := <-watcher.Events:
+			// check if the delay between events is more than a second; if not don't push again
+			if time.Now().After(lastChange.Add(delay)) {
+				lastChange = time.Now()
```

Comment:
@feloy one of the problems I am facing after removing that infinite for loop and doing everything in this single go routine is that https://github.com/redhat-developer/odo/issues/5532 pops up frequently. 

I added the above logic to ensure that we don't do odo dev too often, but only if the gap between two successive events is more than 1 second. Even though it fixed the problem initially, it came back, and I am seeing odo dev push things not just twice but sometimes as much as 4 times!